### ISSUE TITLE
Reentrancy and thread-safety?

### ISSUE THREAD

**Issue by audiamus** [Zeugma440/atldotnet#70]
## The problem

Strange exceptions in conjunction with multi-threaded, parallel access to ATL.net 

## Environment

* ATL.net 3.9
* Win7, .Net 4.7.1

## Details

My tool, AAX Audio Converter, runs much of its conversion task in parallel. As I am extending the usage of ATL.net, I have encountered a number of odd exceptions thrown from ATL.net, e.g. null-pointers.   

It may have to do with instantiating `ATL.Track`: `var tagFile = new ATL.Track (_track.FileName);`

Could it be that this invocation is not reentrant or subsequent usage of the instance not thread-safe?
  
Also, as the `Settings` class is static, there may be other pitfalls for me here as well. 

My bypass for the moment: Synchronizing access with `lock () {}`. That seems to prevent the mishaps.

## Code To Reproduce Issue

Difficult, as these errors could be race conditions. But it's more of a question anyway.

**Comment by Zeugma440** [comment]
Hi and thanks for your feedback. At this time there has been no real testing of the library in a multi-threaded environment. I'll set up some tests and tell you how it goes. 

If anything, I suspect a mis-use of some .NET components rather than actual coding issues. 

The `Settings` class is a quick and dirty implementation and could certainly be improved, for sure. I'll see what I can do on this side as well.

**Comment by Zeugma440** [comment]
@audiamus just to be sure : could you submit the code you're running on a threaded environment ?

I've just done some tests with reading files and it doesn't seem to crash at all, even with 40 concurrent reads.

**Comment by audiamus** [comment]
>could you submit the code you're running on a threaded environment ?

The code is basically the same as last week, now just with the lock added. I had this running without the lock, threads from a `Parallel.ForEach()` loop, on an Intel I7 processor with hyper threading, so up to 8 threads would run in parallel. I was debugging elsewhere in the code, hit a breakpoint in one of the parallel threads, and when continuing got the exception from the `ATL.Track` ctor, possibly in some other thread. The breakpoint hit may play a role here, but shouldn't, of course. 

```
      lock (__lockableATL) {
        ATL.Track tagFile;
        try {

          tagFile = new ATL.Track (_track.FileName);

          ATL.Settings.ID3v2_tagSubVersion = 3;
          ATL.Settings.MP4_createNeroChapters = true;
          ATL.Settings.MP4_createQuicktimeChapters = true;

          tagFile.Chapters.Clear ();

          foreach (var ch in chapters) {
            var chi = new ATL.ChapterInfo {
              StartTime = (uint)ch.Time.Begin.TotalMilliseconds,
              EndTime = (uint)ch.Time.End.TotalMilliseconds,
              Title = ch.Name
            };
            tagFile.Chapters.Add (chi);
          }

          tagFile.Save ();

        } catch (Exception exc) {
          Log (1, this, exc.ToShortString ());
          return false;
        }
      }
```

**Comment by Zeugma440** [comment]
If it has something to do with the constructor, maybe could you try with 

```csharp
tagFile = new ATL.Track (string.Copy(_track.FileName));
```

instead of

```csharp
 tagFile = new ATL.Track (_track.FileName);
```

So far ATL has been using the provided string as is (i.e. by reference). Thus, if you reuse the same string instance in your loop, it might create weird stuff indeed in the library. Let me know if that helps.

NB : I'll fix it on my end as well.

**Comment by audiamus** [comment]
I will try to reproduce it and also try a copy, but it will have to wait a few days. _(This project is a spare time affair.)_

The way it is used, that `FileName `property is immutable. There will a new instance of the encapsulating class for each invocation of this piece of code.

**Comment by Zeugma440** [comment]
Just ran the same code as yours on 4 MP4 files simultaneously (4 threads) without any issue.

Same behaviour with `Parallel.ForEach` and with manual thread management `new Thread` / `Thread.Start`.

=> I'll need more feedback from you about how to reproduce it, and which stacktrace you get.

Last but not least, on the code you posted, `ATL.Settings` may be set once every run. The `Settings` class is no more than a static container, there's no need to set the settings upon every loop iteration.

Oh, and don't worry about being reactive, ATL is a spare time affair as well :wink:

**Comment by audiamus** [comment]
Well then. It's quite strange. I managed to reproduce an exception, but there isn't much of a call stack. (It only fired in Release mode but that wouldn't matter for ATL.net being the assembly from the NuGet package.)

And, the lock didn't actually help. Same exception, invocation for some other track and from a different thread. 

![exception](https://user-images.githubusercontent.com/47717419/93373009-05b6dc00-f855-11ea-8ebd-0c1a3c5b1300.png)

```
System.ArgumentException
  HResult=0x80070057
  Message=An item with the same key has already been added.
  Source=mscorlib
  StackTrace:
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
```

**Comment by audiamus** [comment]
More info.

I downloaded your source and built ATL.net myself, instead of using the NuGet package. This allows us to inspect the inner workings of ATL.net.

And now I get the null pointer exception again with this call stack:

```
    System.Collections.Generic.Dictionary<string, System.Collections.Generic.IList<ATL.Format>>.Insert(string key = ".mp1", System.Collections.Generic.IList<ATL.Format> value = Count = 1, bool add)
 	System.Collections.Generic.Dictionary<System.__Canon, System.__Canon>.Add(System.__Canon key, System.__Canon value)
>	ATL.Factory.addFormat(ATL.Format f = {ATL.Format}) Line 38
 	ATL.AudioData.AudioDataIOFactory.GetInstance() Line 82
 	ATL.AudioData.AudioFileIO.AudioFileIO(string path = "\\\\?\\F:\\Temp\\aax\\Book Author\\Book Title\\03 - Book Title - Book Author.m4b", bool readEmbeddedPictures = false, bool readAllMetaFrames = true, System.IProgress<float> writeProgress = null) Line 35
 	ATL.Track.Update(bool onlyReadEmbeddedPictures = false) Line 272
 	ATL.Track.Track(string Path = "\\\\?\\F:\\Temp\\aax\\Book Author\\Book Title\\03 - Book Title - Book Author.m4b", System.IProgress<float> writeProgress = null, bool load = true) Line 31
 	audiamus.aaxconv.lib.TagAndFileNamingHelper.writeMetaDataChapters() Line 74
```

**Comment by Zeugma440** [comment]
Awesome. Thanks a lot.

Might be my Factory pattern that has an implementation issue. I'll check it out and keep you informed!

**Comment by Zeugma440** [comment]
Latest commit (42a7e5d9d43ce549a96a9c3d877eb0c5026da649) should solve that issue.

Could you grab it, rebuild ATL and see if it works ?

**Comment by audiamus** [comment]
Thank you very much, I will test it. It has to wait til tomorrow or the weekend, though. Difficulty is that it was still hard to reproduce. I have been "lucky" only two or three times yesterday, among dozens of attempts. 

I noticed you use `[MethodImpl(MethodImplOptions.Synchronized)]` in your fix. As this is applied to a public type it increases the risk of deadlocks, see this old blog entry:
https://docs.microsoft.com/en-us/archive/blogs/bclteam/lock-vs-methodimploptions-synchronized-kit-george

I always do something like this:
```
  public class Foo {
    private readonly object _lockable = new object ();

    public void Bar () {
      lock (_lockable) {
        // synchronized code
      }
    }
  }
```

_Re: the static Settings class_:
Readjusting the settings on every "Chapters" invocation is just a safety catch for me. It ensures that they will have the correct values as they could have easily been altered elsewhere in the application code.

**Comment by Zeugma440** [comment]
I see, thanks for pointing to this.

The bug-free solution is less elegant but does ensure nothing will go wrong if someone uses the same object in his lock.

I'll adjust it very soon.

**Comment by audiamus** [comment]
I have now tested the latest commit, with my locking disabled, with a number of different books and several runs per book. **No exception was thrown.** It's not exhaustive, of course. On the other hand: Creating the instance of your factory singleton clearly wasn't thread-safe. But as it is now, I guess problem is solved.   

_My own lock didn't actually work at first. While I named it_ `__lockableATL` _with the double underscore, standing for a static field in my notation, it wasn't static but an instance field. I create a new instance of this class for every track to be post-processed, so my locking attempt did not have any effect at all. But no longer needed now, I guess._  

### Unrelated:
_I encountered another phenomenon, probably harmless as it only occurs with **Debug** builds, not with **Release** builds:
During_ `Track.Save ()`_, memory usage rises dramatically, from below 100 MB to 1.6 or 2 GB and it will never be freed again. This might be unmanaged memory. That's what I guess from Visual Studio Diagnostics, as managed memory only marginally increases._  

![memusage](https://user-images.githubusercontent.com/47717419/93664199-69b3ed00-fa6d-11ea-8e7d-b61f2eaed132.png)

**Comment by Zeugma440** [comment]
Hi and thanks for your feedback. I'll change the implementation of my lock to reduce risks and publish a new version under NuGet very soon.

Regarding memory usage, my first guess would be that the .NET file I/O environment is not very optimized when used in debug mode. Plus if I'm not mistaken, you're working with chapters on MP4 audiobooks. MP4s have a very complex structure that requires ATL to navigate the entire file to properly adjust metadata (more info [here](https://github.com/Zeugma440/atldotnet/wiki/Focus-on-Chapter-metadata#mp4)). The memory usage you're witnessing might be the consequence of navigating the whole file (audiobook => 100's of MBs) while editing its chapters.

Last but not least, I can assure you that the library never keeps raw file data in memory unless it's essential for processing. While editing metadata, ATL loads _chunks of files_, rewrites them accordingly, and loads the next chunk to do the same. It never loads the entire file unless you change default settings to make it do so.

**Comment by Zeugma440** [comment]
Fix published in v3.11
